File: app/src/main/java/in/androidmate/anujgupta/notes_room/database/LocalCacheManager.java
package in.androidmate.anujgupta.notes_room.database;

import android.arch.persistence.room.Room;
import android.content.Context;

import java.util.List;

import in.androidmate.anujgupta.notes_room.models.Note;
import in.androidmate.anujgupta.notes_room.ui.add_note.AddNoteViewInterface;
import in.androidmate.anujgupta.notes_room.ui.main.MainViewInterface;
import io.reactivex.Completable;
import io.reactivex.CompletableObserver;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Action;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by anujgupta on 14/01/18.
 */

public class LocalCacheManager {
    private Context context;
    private static LocalCacheManager _instance;
    private AppDatabase db;

    public static LocalCacheManager getInstance(Context context) {
        if (_instance == null) {
            _instance = new LocalCacheManager(context);
        }
        return _instance;
    }

    public LocalCacheManager(Context context) {
        this.context = context;
        db = AppDatabase.getAppDatabase(context);
    }

    public void getNotes(final MainViewInterface mainViewInterface) {
        db.noteDao().getAll().subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread()).subscribe(new Consumer<List<Note>>() {
            @Override
            public void accept(List<Note> notes) throws Exception {
                mainViewInterface.onNotesLoaded(notes);
            }
        });
    }

    public void addNotes(final AddNoteViewInterface addNoteViewInterface, final String title, final String note_text) {
        Completable.fromAction(new Action() {
            @Override
            public void run() throws Exception {
                Note note = new Note(title, note_text);
                db.noteDao().insertAll(note);
            }
        }).observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io()).subscribe(new CompletableObserver() {
            @Override
            public void onSubscribe(Disposable d) {
            }

            @Override
            public void onComplete() {
                addNoteViewInterface.onNoteAdded();
            }

            @Override
            public void onError(Throwable e) {
                addNoteViewInterface.onDataNotAvailable();
            }
        });
    }


}

File: app/src/main/java/dhbw/androidmate/anujgupta/notes_room/database/LocalCacheManager.kt
package dhbw.androidmate.anujgupta.notes_room.database

import kotlin.Throws
import dhbw.androidmate.anujgupta.notes_room.ui.main.MainViewInterface
import dhbw.androidmate.anujgupta.notes_room.R
import dhbw.androidmate.anujgupta.notes_room.ui.add_note.AddNoteActivity
import dhbw.androidmate.anujgupta.notes_room.adapters.NotesAdapter
import dhbw.androidmate.anujgupta.notes_room.ui.add_note.AddNoteViewInterface
import dhbw.androidmate.anujgupta.notes_room.ui.main.MainActivity
import dhbw.androidmate.anujgupta.notes_room.adapters.NotesAdapter.NotesViewHolder
import dhbw.androidmate.anujgupta.notes_room.database.NoteDao
import dhbw.androidmate.anujgupta.notes_room.database.AppDatabase
import dhbw.androidmate.anujgupta.notes_room.models.Note
import android.content.Context
import io.reactivex.schedulers.Schedulers
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.Completable
import io.reactivex.CompletableObserver
import io.reactivex.disposables.Disposable
import dhbw.androidmate.anujgupta.notes_room.database.LocalCacheManager
import io.reactivex.functions.Consumer
import java.lang.Exception

class LocalCacheManager(private val context: Context?) {
    private val db: AppDatabase?

    fun getNotes(mainViewInterface: MainViewInterface?) {
        db?.noteDao()?.getAll()?.subscribeOn(Schedulers.io())?.observeOn(AndroidSchedulers.mainThread())?.subscribe(object : Consumer<MutableList<Note?>?> {
            @Throws(Exception::class)
            override fun accept(t: MutableList<Note?>) {
                mainViewInterface?.onNotesLoaded(t)
            }
        })
    }

    fun addNotes(addNoteViewInterface: AddNoteViewInterface?, title: String?, note_text: String?) {
        Completable.fromAction {
            val note = Note(title, note_text)
            db?.noteDao()?.insertAll(note)
        }.observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io()).subscribe(object : CompletableObserver {
                    override fun onSubscribe(d: Disposable?) {}
                    override fun onComplete() {
                        addNoteViewInterface?.onNoteAdded()
                    }

                    override fun onError(e: Throwable?) {
                        addNoteViewInterface?.onDataNotAvailable()
                    }
                })
    }

    companion object {
        private var _instance: LocalCacheManager? = null
        fun getInstance(context: Context?): LocalCacheManager? {
            if (_instance == null) {
                _instance = LocalCacheManager(context)
            }
            return _instance
        }
    }

    init {
        db = AppDatabase.Companion.getAppDatabase(context)
    }
}